server/controllers/CommentController.js

```diff
@@ -164,26 +168,77 @@ exports.reply = async (req, res) => {
   // This is a reply to an AI generated comment.
 
   console.log("Generating AI reply...");
-  const comments = formatComments(
-    await Comment.findAll({
-      where: {
-        solutionId: parent.solutionId,
-      },
+
+  const fetchedComments = await Comment.findAll({
+    where: {
+      solutionId: parent.solutionId,
+    },
+  });
+
+  const extracted = fetchedComments.map(
+    ({ id, text, userId, solutionId, upVotes, replies, usersWhoUpvoted }) => ({
+      id,
+      text,
+      userId,
+      solutionId,
+      hasUserUpvoted: usersWhoUpvoted.includes(req.user.username),
+      upVotes,
+      replies,
     }),
   );
 
+  const comments = formatComments(extracted);
+
+  // console.log(comments);
+
   const comment_chain = comments
     .map((c) => getCommentChain(c, comment.id))
     .filter((ch) => ch.length !== 0)[0];
   const solution = await Solution.findByPk(parent.solutionId);
   const challenge = await Challenge.findByPk(solution.challengeId);
 
-  console.log("Calling AI TA....");
-  const feedback = await AITA.feedback_for_comment(
+  console.log("Calling AITA....");
+  const [chainRunId, feedback] = await AITA.feedback_for_comment(
     comment_chain,
     challenge,
     solution,
   );
-  await reply_to_comment(comment.id, feedback, "AITA");
-  console.log(`AI TA commented on comment ${comment.id}`);
+
+  await reply_to_comment(comment.id, feedback, "AITA", chainRunId);
+  console.log(`AITA commented on comment ${comment.id}`);
+};
+
+exports.upvote = async (req, res) => {
+  const { commentId } = req.params;
+
+  const comment = await Comment.findByPk(commentId);
+
+  const upvotedUsersList = comment.usersWhoUpvoted
+    .split(",")
+    .filter((v) => v.length !== 0);
+
+  if (upvotedUsersList.includes(`${req.user.username}`)) {
+    console.log("Invalid upvote.");
+    res.status(409).send();
+    return;
+  }
+
+  Comment.update(
+    {
+      upVotes: comment.upVotes + 1,
+      usersWhoUpvoted: [...upvotedUsersList, `${req.user.username}`].join(","),
+    },
+    {
+      where: {
+        id: comment.id,
+      },
+    },
+  );
+
+  res.status(204).send();
+
+  if (comment.userId === "AITA") {
```

Comment:
Is the AI like a regular user in our database? it maybe good to have a separate field for ai comments but if the AI is a user then I think we can just sort the comments by userid of the ai

Reply:
Yes we are treating the AI as a user